### PR TITLE
Auto opt-in DatabricksSDKModelsArtifactRepository for SEG enabled databricks workspaces.

### PR DESCRIPTION
**Pull request by shichengzhou-db**
### Related Issues/PRs
https://github.com/mlflow/mlflow/pull/13203

### What changes are proposed in this pull request?
In [PR 13203](https://github.com/mlflow/mlflow/pull/13203), we introduced a DatabricksSDKModelsArtifactRepository that leverage Databricks SDK for UC model registry, and allow opting in to use it via an env var.

In this PR, we will
- Automatically opt-in eligible databricks workspace(SEG enabled, Internet-Off workspaces) to use DatabricksSDKModelsArtifactRepository.
- Emit model version lineage explicitly.
  - For context, we used to emit model version lineage in a call to `GenerateTemporaryCredential` in databricks control plane. That call will not exist when we use DatabricksSDKModelsArtifactRepository. Therefore, we are explicitly emitting the lineage through calling a new endpoint in databricks control plane.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [x] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?
`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/utils/_unity_catalog_utils.py [hunk 1]
```diff
@@ -374,3 +393,57 @@ def get_full_name_from_sc(name, spark) -> str:
...
+            json_body=req_body,
+            response_proto=response_proto,
+        )
+        is_enabled = resp.is_databricks_sdk_models_artifact_repository_enabled
+    except Exception as e:
+        eprint(
+            "Failed to confirm if DatabricksSDKModelsArtifactRepository should be used; "
+            f"falling back to default. Error: {e}"
+        )
+    return is_enabled
```
Comment (harupy) [inline mlflow/utils/_unity_catalog_utils.py:423]:
```suggestion
    try:
        resp = call_endpoint(
            host_creds=host_creds,
            endpoint=endpoint,
            method=method,
            json_body=req_body,
            response_proto=response_proto,
        )
        return resp.is_databricks_sdk_models_artifact_repository_enabled
    except Exception as e:
        eprint(
            "Failed to confirm if DatabricksSDKModelsArtifactRepository should be used; "
            f"falling back to default. Error: {e}"
        )
    return False
```

File: mlflow/utils/_unity_catalog_utils.py [hunk 2]
```diff
@@ -374,3 +393,57 @@ def get_full_name_from_sc(name, spark) -> str:
...
+        resp = call_endpoint(
+            host_creds=host_creds,
+            endpoint=endpoint,
+            method=method,
+            json_body=req_body,
+            response_proto=response_proto,
+        )
+        is_enabled = resp.is_databricks_sdk_models_artifact_repository_enabled
+    except Exception as e:
+        eprint(
```
Comment (harupy) [inline mlflow/utils/_unity_catalog_utils.py:419]:
```suggestion
        _logger.warning(
```

Can we use `_logger.warning` instead?

File: tests/store/artifact/test_databricks_sdk_models_artifact_repo.py [hunk 3]
```diff
@@ -173,19 +174,59 @@ def test_mlflow_use_databricks_sdk_model_artifacts_repo_for_uc(tmp_path, monkeyp
...
+        }
     )
+    with mock.patch(
+        "mlflow.utils._unity_catalog_utils.call_endpoint",
+    ) as mock_call_endpoint:
+        mock_call_endpoint.side_effect = [
+            SimpleNamespace(is_databricks_sdk_models_artifact_repository_enabled=True),
+            Exception("lineage emission fails"),
+            SimpleNamespace(is_databricks_sdk_models_artifact_repository_enabled=True),
+        ]
```
Comment (harupy) [inline tests/store/artifact/test_databricks_sdk_models_artifact_repo.py:215]:
```suggestion
    with mock.patch(
        "mlflow.utils._unity_catalog_utils.call_endpoint",
        side_effect=[
            SimpleNamespace(is_databricks_sdk_models_artifact_repository_enabled=True),
            Exception("lineage emission fails"),
            SimpleNamespace(is_databricks_sdk_models_artifact_repository_enabled=True),
        ]
    ) as mock_call_endpoint:
```

File: mlflow/utils/_unity_catalog_utils.py [hunk 4]
```diff
@@ -374,3 +393,57 @@ def get_full_name_from_sc(name, spark) -> str:
...
+            host_creds=host_creds,
+            endpoint=endpoint,
+            method=method,
+            json_body=req_body,
+            response_proto=response_proto,
+        )
+    except Exception as e:
+        eprint(
+            f"Failed to emit best-effort model version lineage. Error: {e}"
+        )
```
[the remark this answers is not in the document]
Comment (shichengzhou-db) [reply mlflow/utils/_unity_catalog_utils.py:449]: Yeah, we provide best-effort model version lineage, but in 99.99% of the cases we want this to work, and I am hoping in rare/unexpected cases when this fail,  customers / field engs can have a better understanding of what's going on for trouble-shooting.

**Review by harupy — APPROVED** [review]

**Comment by harupy** [comment]
Can we address https://github.com/mlflow/mlflow/pull/13632#discussion_r1832220771?

File: mlflow/utils/_unity_catalog_utils.py [hunk 5]
```diff
@@ -374,3 +390,55 @@ def get_full_name_from_sc(name, spark) -> str:
...
+            json_body=req_body,
+            response_proto=response_proto,
+        )
+        return resp.is_databricks_sdk_models_artifact_repository_enabled
+    except Exception as e:
+        _logger.warning(
+            "Failed to confirm if DatabricksSDKModelsArtifactRepository should be used; "
+            f"falling back to default. Error: {e}"
+        )
+        return False
```
Comment (harupy) [inline mlflow/utils/_unity_catalog_utils.py:418]:
```suggestion

    return False
```

File: mlflow/utils/_unity_catalog_utils.py [hunk 6]
```diff
@@ -374,3 +390,55 @@ def get_full_name_from_sc(name, spark) -> str:
         return f"{catalog}.{name}"
     schema = spark.sql(_ACTIVE_SCHEMA_QUERY).collect()[0]["schema"]
     return f"{catalog}.{schema}.{name}"
+
+
+def is_databricks_sdk_models_artifact_repository_enabled(host_creds):
+    # Return early if the environment variable is set to use the SDK models artifact repository
+    if MLFLOW_USE_DATABRICKS_SDK_MODEL_ARTIFACTS_REPO_FOR_UC.defined:
+        return MLFLOW_USE_DATABRICKS_SDK_MODEL_ARTIFACTS_REPO_FOR_UC.get()
```
Comment (harupy) [inline mlflow/utils/_unity_catalog_utils.py:398]:
```suggestion
        return True
```

Comment (shichengzhou-db) [reply mlflow/utils/_unity_catalog_utils.py:398]: @harupy if user overrides this to be false explicitly, returning True is incorrect right?

Comment (harupy) [reply mlflow/utils/_unity_catalog_utils.py:398]: ah makes sense!

**Pull request merged**